### ISSUE TITLE
express-handlebars.d.ts has missing types

### ISSUE THREAD

**Issue by sr258** [express-handlebars/express-handlebars#243]
When I use this library with the TypeScript transpiler (4.5.2) I get the following errors:

```
node_modules/express-handlebars/dist/express-handlebars.d.ts:13:22 - error TS2304: Cannot find name 'CompileOptions'.

13     compilerOptions: CompileOptions;
                        ~~~~~~~~~~~~~~

node_modules/express-handlebars/dist/express-handlebars.d.ts:14:21 - error TS2304: Cannot find name 'RuntimeOptions'.

14     runtimeOptions: RuntimeOptions;
                       ~~~~~~~~~~~~~~

node_modules/express-handlebars/dist/express-handlebars.d.ts:20:78 - error TS2304: Cannot find name 'HandlebarsTemplateDelegate'.

20     getTemplate(filePath: string, options?: PartialTemplateOptions): Promise<HandlebarsTemplateDelegate | TemplateSpecification>;
                                                                                ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/express-handlebars/dist/express-handlebars.d.ts:20:107 - error TS2304: Cannot find name 'TemplateSpecification'.

20     getTemplate(filePath: string, options?: PartialTemplateOptions): Promise<HandlebarsTemplateDelegate | TemplateSpecification>;
                                                                                                             ~~~~~~~~~~~~~~~~~~~~~

node_modules/express-handlebars/dist/express-handlebars.d.ts:21:78 - error TS2304: Cannot find name 'HandlebarsTemplateDelegate'.

21     getTemplates(dirPath: string, options?: PartialTemplateOptions): Promise<HandlebarsTemplateDelegate | TemplateSpecification>;
                                                                                ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/express-handlebars/dist/express-handlebars.d.ts:21:107 - error TS2304: Cannot find name 'TemplateSpecification'.

21     getTemplates(dirPath: string, options?: PartialTemplateOptions): Promise<HandlebarsTemplateDelegate | TemplateSpecification>;
                                                                                                             ~~~~~~~~~~~~~~~~~~~~~

node_modules/express-handlebars/dist/express-handlebars.d.ts:27:60 - error TS2304: Cannot find name 'RuntimeOptions'.

27     protected _compileTemplate(template: string, options?: RuntimeOptions): HandlebarsTemplateDelegate;
                                                              ~~~~~~~~~~~~~~

node_modules/express-handlebars/dist/express-handlebars.d.ts:27:77 - error TS2304: Cannot find name 'HandlebarsTemplateDelegate'.

27     protected _compileTemplate(template: string, options?: RuntimeOptions): HandlebarsTemplateDelegate;
                                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/express-handlebars/dist/express-handlebars.d.ts:28:63 - error TS2304: Cannot find name 'RuntimeOptions'.

28     protected _precompileTemplate(template: string, options?: RuntimeOptions): TemplateSpecification;
                                                                 ~~~~~~~~~~~~~~

node_modules/express-handlebars/dist/express-handlebars.d.ts:28:80 - error TS2304: Cannot find name 'TemplateSpecification'.

28     protected _precompileTemplate(template: string, options?: RuntimeOptions): TemplateSpecification;
                                                                                  ~~~~~~~~~~~~~~~~~~~~~

node_modules/express-handlebars/dist/express-handlebars.d.ts:29:41 - error TS2304: Cannot find name 'HandlebarsTemplateDelegate'.

29     protected _renderTemplate(template: HandlebarsTemplateDelegate, context?: UnknownObject, options?: RuntimeOptions): string;
                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/express-handlebars/dist/express-handlebars.d.ts:29:104 - error TS2304: Cannot find name 'RuntimeOptions'.

29     protected _renderTemplate(template: HandlebarsTemplateDelegate, context?: UnknownObject, options?: RuntimeOptions): string;
                                                                                                          ~~~~~~~~~~~~~~

node_modules/express-handlebars/types/index.d.ts:10:19 - error TS2503: Cannot find namespace 'Handlebars'.

10  [index: string]: Handlebars.TemplateDelegate;
                     ~~~~~~~~~~

node_modules/express-handlebars/types/index.d.ts:14:19 - error TS2304: Cannot find name 'TemplateSpecification'.

14  [index: string]: TemplateSpecification;
                     ~~~~~~~~~~~~~~~~~~~~~

node_modules/express-handlebars/types/index.d.ts:18:27 - error TS2503: Cannot find namespace 'Handlebars'.

18  [index: string]: Promise<Handlebars.TemplateDelegate>;
                             ~~~~~~~~~~

node_modules/express-handlebars/types/index.d.ts:22:27 - error TS2304: Cannot find name 'TemplateSpecification'.

22  [index: string]: Promise<TemplateSpecification>;
                             ~~~~~~~~~~~~~~~~~~~~~

node_modules/express-handlebars/types/index.d.ts:51:19 - error TS2503: Cannot find namespace 'Handlebars'.

51  runtimeOptions?: Handlebars.RuntimeOptions;
                     ~~~~~~~~~~

node_modules/express-handlebars/types/index.d.ts:61:59 - error TS2304: Cannot find name 'CompileOptions'.

61 export type HandlebarsCompile = (input: unknown, options: CompileOptions) => Handlebars.TemplateDelegate;
                                                             ~~~~~~~~~~~~~~

node_modules/express-handlebars/types/index.d.ts:61:78 - error TS2503: Cannot find namespace 'Handlebars'.

61 export type HandlebarsCompile = (input: unknown, options: CompileOptions) => Handlebars.TemplateDelegate;
                                                                                ~~~~~~~~~~

node_modules/express-handlebars/types/index.d.ts:62:62 - error TS2304: Cannot find name 'PrecompileOptions'.

62 export type HandlebarsPrecompile = (input: unknown, options: PrecompileOptions) => TemplateSpecification;
                                                                ~~~~~~~~~~~~~~~~~

node_modules/express-handlebars/types/index.d.ts:62:84 - error TS2304: Cannot find name 'TemplateSpecification'.

62 export type HandlebarsPrecompile = (input: unknown, options: PrecompileOptions) => TemplateSpecification;
                                                                                      ~~~~~~~~~~~~~~~~~~~~~

node_modules/express-handlebars/types/index.d.ts:78:20 - error TS2304: Cannot find name 'CompileOptions'.

78  compilerOptions?: CompileOptions;
                      ~~~~~~~~~~~~~~

node_modules/express-handlebars/types/index.d.ts:79:19 - error TS2503: Cannot find namespace 'Handlebars'.

79  runtimeOptions?: Handlebars.RuntimeOptions;
                     ~~~~~~~~~~
```

I've looked at express-handlebars.d.ts and the TypeScript compiler seems correct. The names are not imported, so it can't find them.

**Comment by UziTech** [comment]
Those types are from handlebars

**Comment by sr258** [comment]
Yes, but they are not imported by the .d.ts file, so TypeScript can't resolve them.

**Comment by UziTech** [comment]
PRs are welcome 😁👍

**Comment by sr258** [comment]
I've tried fixing this, but the problem isn't as easy to solve as a though:

The handlebars package doesn't export the interfaces that cause the errors, so TypeScript can't reference them in the .d.t.s file. If I add a manual reference to Handlebars in the .d.ts file things work: `import * as Handlebars from 'handlebars';`

I don't understand why importing all exported members from the handlebar module even works at all in your code. As CompileOptions is not exported by the handlebars type definition, it should be usable in  express-handlebars.ts. Did this work out of the box for you or did you do something special so that TypeScript knows about the definitions for CompileOptions and alike?

**Comment by UziTech** [comment]
I'm no Typescript expert but I think the way handlebars "declare"s the types (instead of exporting) they should be in the global scope and not need to be imported as long as handlebars is in your dependencies.

**Comment by UziTech** [comment]
or maybe we need to reference them some way?
https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html

**Comment by UziTech** [comment]
try adding `/// <reference types="handlebars" />` to the top of the .d.ts file in express-handlebars and see if that works.

**Comment by UziTech** [comment]
Fixed in #247
